contracts/forwarder/Eip712Forwarder.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier:MIT
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "./IForwarder.sol";
+// TODO now: move getError to the forwarder folder
+import "../utils/GsnUtils.sol";
+
+contract Eip712Forwarder is IForwarder {
+    using ECDSA for bytes32;
+
+    string public constant GENERIC_PARAMS = "address to,bytes data,address from,uint256 nonce,uint256 gas";
+
+    mapping(bytes32 => bool) public typeHashes;
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    function getNonce(address from)
+    public view override
+    returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig)
+    public override view {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+    }
+
+    function execute(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig
+    )
+    public
+    override
+    returns (bool, bytes memory) {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        return req.to.call{gas : req.gas}(abi.encodePacked(req.data, req.from));
+    }
+
+    function _verifyNonce(ForwardRequest memory req) internal view {
+        require(nonces[req.from] == req.nonce, "nonce mismatch");
```

Comment:
Not planning to support different nonce strategy (provided by signer, like https://github.com/wighawag/eip-2585/blob/c9769ad0b4260b4f08f981527bb1162ecce570d3/contracts/src/EIP712Forwarder.sol#L50) ?

Reply:
Its quite possible to add such nonce policy.
We're leaving it to a separate pull request, as this one is already too big (and also it doesn't require an API change)
The nice thing adding batch/nonce can be done not only without solidity API change, but even without client change: the standard logic of latest-nonce-plus-1 is good even if the current batch was changed.